### ISSUE TITLE
Dynamic room descriptions

### ISSUE THREAD

**Issue by irondavy** [ckolderup/picaro#4]
I'm not sure if this is a function we need but we might want to consider how room descriptions are generated. If an item changes state, or is removed/destroyed, the room description might need to update to reflect that. Naturally, if we just say that the room state is fixed, people will write it accordingly.

It's hard to imagine, though, how you'd tell a player that there's a takable item in the room in a way that would still make sense after the item has been taken. I see three solutions, but perhaps there's something I'm not considering:
- Room descriptions are conditional based on room state, or maybe counter? More work & more complex.
- Room descriptions include descriptions of all items in the room. Kinda messy.
- Room descriptions include a flat list of all items in the room. Kinda boring.

In the latter two cases, I'm guessing we'll need some way of saying the item is immediately known to the player or not. Having the player explore for items seems important, especially since we don't currently have nested items (right?).

**Comment by ckolderup** [comment]
I have some notes on this from my more complicated, all-encompassing
attempt at defining stuff (which may be eventually worth revisiting as
a file format v2) that I'll try to dig up— both on generating room
descriptions and having it possible for items to be "visible" or not.

On Wed, Jan 25, 2012 at 11:52 AM, David Cole
reply@reply.github.com
wrote:

**Comment by hoverbird** [comment]
@irondavy option #3, while boring (or at least vanilla IF), does give the player everything they need without requiring extra work on the part of the creator.  I could see there being a general-purpose facility for changing descriptions at some point, which could be triggered on top of this, but that doesn't seem important for v1.

**Comment by irondavy** [comment]
I'm inclined to agree, for V1. How do we print it?

```
[Room description]
You see: [item], [item], [item].
```

Related: can we generate articles? Does this make sense: do "a" and "an" for everything except for items with capitalized names which would get no article. This would also be useful for the verb modals where the language is a little stilted (_Take bat_ instead of _Take the bat_).

**Comment by ckolderup** [comment]
Articles are totally doable. If the name starts with a capital letter or "the " (with a space), don't add it, otherwise check for vowel, etc

Your format seems fine. Maybe if there's only one item it just says `You see [item] here.` ?

**Comment by irondavy** [comment]
Cool, I figured that was a solved problem from IF but I didn't know what it would be.

And to be clear, we're explicitly leaving out any "You enter the [room name]" type message.

**Comment by ckolderup** [comment]
re: entering room messages: yeah, I think so. We have a persistent and fairly bold current-room label in the header.

Also, I looked up what I was originally thinking for visible vs "hidden" items. It seemed like the "easiest" solution (maybe not the cleanest) was to support a custom room-is-describing-you string, which you could override. This would let you make the object "invisible" as well as letting you put in custom descriptions, in case you wanted to add some suspense— the "amulet" has custom room-describing-you string "faint shimmering", then you'd look at the shimmering to get the amulet? Something like that. If the string was empty, we would just not insert an extra line break in the list of items on room entrance or in the item picker. (Obviously we'd need some way for that to change, though, so it would show up in your inventory...)

**Comment by irondavy** [comment]
That's interesting! I'd be curious to hear what @JoeBernardi and @hoverbird think about that proposal.

**Comment by hoverbird** [comment]
I think room descriptions being static is a really helpful V1 constraint. You could get a similiar effect if you just had an item called "Faint shimmering" in the room.  When you enter, you'd get the room description + "you see a faint shimmering here".  

Looking at it fired a "replace this with an amulet" event and a "put this in my inventory event".  It's not great literature, but it is very IF-ey and very simple, which is nice for V1.

**Comment by ckolderup** [comment]
Oh, sure, if we're going forward with the event thing, it seems easy to just do a "put the amulet in my inventory" event and a "hide/destroy/fling myself into the void" event. I just wasn't sure where that stood.

**Comment by hoverbird** [comment]
Same here, I don't think we've decided on it.  I'm suggesting events because the meat of the game creator and player is written in JS, which is generally written in an event-driven style. If you look at the runner code so far, for instance, it's all basically jQuery click event handlers.  What I'm proposing is doing more or less everything that way- in jQuery it's as easy to trigger a "itemAddedToInventory" event as it is to trigger "click".

**Comment by irondavy** [comment]
Yeah, that makes a lot of sense actually. I'm content with this. I'll leave it open to garner any concerns or objections, but if everyone is okay with it then I'll close and call this the spec for V1 room descriptions.

@ardubs @JoeBernardi @ckolderup

**Comment by ckolderup** [comment]
:+1:

**Comment by ardubs** [comment]
the only thing i'd add is that auto-displaying items should be optional -- in the case of the orchard (and i would imagine other cases where there are more than say five objects in a room) you'd want to avoid listing everything in favor of a more general "there are generally yellow apples here." then the item picker itself can show you exactly what's there.

**Comment by irondavy** [comment]
I _really_ want to avoid adding options, as we've mostly avoided them entirely. Is there a way to do what you want with a technique like @hoverbird's comment:

> I think room descriptions being static is a really helpful V1 constraint. You could get a similiar effect if you just had an item called "Faint shimmering" in the room. When you enter, you'd get the room description + "you see a faint shimmering here".
> 
> Looking at it fired a "replace this with an amulet" event and a "put this in my inventory event". It's not great literature, but it is very IF-ey and very simple, which is nice for V1.

**Comment by ardubs** [comment]
Well an event-driven swap feature sounds like it could be pretty useful, and I like that it's not just a rehash of how IF does things. We've discussed using lists within JSON to encapsulate multiple item states that in turn can represent different levels of player knowledge:

"description" : ["a faint shimmering", "a magical amulet", "the dread Amulet of Klarb"]

...would the event-driven model be a replacement for this? Or would firing a "move to next state" event when the player "looks" at the amulet be a use case for it? Because if so, we can do what I'm talking about apple-wise with:

"description" : ["", "mealy and brown"]

Where the room description text for the apple is empty by default, and only becomes "mealy and brown" once the event of 
having been placed in the player's inventory occurs.

**Comment by ardubs** [comment]
This is sort of an existing issue in IF -- the fact that there are multiple contexts in which an item in the world would need to provide text about itself, each suggesting a different kind of flavor:
- The room description. "A mealy brown apple rots on the ground."
- The player examines the apple, getting more detail. "It's mealy and brown and you don't want to eat it."
- The player checks their inventory. "A mealy brown apple."

I'm curious to know whether we see ourselves as collapsing these down into fewer contexts -- i.e. looking to constrain our world model so that wrinkles like this aren't necessary -- or trying to achieve the same number of contexts, just in a different way.
